src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -1,28 +1,170 @@
+            user.apiKeys
+                .stream()
+                .filter(ApiKey -> apiKeyId.equals(ApiKey.id))
+                .findAny()
+                .orElse(null) != null;
+    }
+
+    /**
+     * Create a new API key and assign it to the provided usage plan. If no usage plan is provided use the default
+     * usage plan instead.
+     */
+    private ApiUser createApiKeyForApiUser(Request req, Response res) {
+        ApiUser targetUser = getApiUser(req);
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        String usagePlanId = req.queryParamOrDefault("usagePlanId", DEFAULT_USAGE_PLAN_ID);
+        // If requester is not an admin user, force the usage plan ID to the default and enforce key limit. A non-admin
+        // user should not be able to create an API key for any usage plan.
+        if (!isUserAdmin(requestingUser)) {
+            usagePlanId = DEFAULT_USAGE_PLAN_ID;
+            if (targetUser.apiKeys.size() >= API_KEY_LIMIT_PER_USER) {
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "User has reached API key limit.");
+            }
+        }
+        // FIXME Should an Api user be limited to one api key per usage plan (and perhaps stage)?
+        CreateApiKeyResult apiKeyResult = ApiGatewayUtils.createApiKey(targetUser.id, usagePlanId);
+        if (apiKeyResult == null || apiKeyResult.getId() == null) {
+            logMessageAndHalt(req,
+                HttpStatus.INTERNAL_SERVER_ERROR_500,
+                String.format("Unable to get AWS api key for user (%s)", targetUser),
+                null
+            );
+            return null;
+        }
+        // Add new API key to user and persist
+        targetUser.apiKeys.add(new ApiKey(apiKeyResult));
+        Persistence.apiUsers.replace(targetUser.id, targetUser);
+        return Persistence.apiUsers.getById(targetUser.id);
+    }
+
+    /**
+     * Delete an api key from a given user's list of api keys (if present) and from AWS api gateway.
+     */
+    private ApiUser deleteApiKeyForApiUser(Request req, Response res) {
+        ApiUser targetUser = getApiUser(req);
+        String apiKeyId = HttpUtils.getRequiredParamFromRequest(req, "apiKeyId", false);
+        if (apiKeyId == null) {
+            logMessageAndHalt(req,
+                HttpStatus.BAD_REQUEST_400,
+                "An api key id is required",
+                null);
+        }
+        if (!userHasKey(targetUser, apiKeyId)) {
+            logMessageAndHalt(req,
+                HttpStatus.NOT_FOUND_404,
+                String.format("User id (%s) does not have expected api key id (%s)", targetUser.id, apiKeyId),
+                null);
+        }
+
+        // Delete api key from AWS.
+        boolean success = ApiGatewayUtils.deleteApiKeys(Collections.singletonList(new ApiKey(apiKeyId)));
+        if (success) {
+            // Delete api key from user and persist
+            targetUser.apiKeys.remove(new ApiKey(apiKeyId));
```

Comment:
Maybe this a pretty standard pattern in Java, but it feels like a more reliable way to do this would be to just call `removeIf`:

```
targetUser.apiKeys.removeIf(apiKey -> apiKeyId.equals(apiKey.id));
```

That would allow for the removal of `ApiKey#equals` and `ApiKey#hashCode`. Open to being overruled if anyone feels strongly about this.

Reply:
This is way cleaner. Have updated.